Neo4j.Driver/Neo4j.Driver/Internal/Routing/LeastConnectedLoadBalancingStrategy.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Copyright (c) 2002-2017 "Neo Technology,"
+// Network Engine for Objects in Lund AB [http://neotechnology.com]
+// 
+// This file is part of Neo4j.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using Neo4j.Driver.V1;
+
+namespace Neo4j.Driver.Internal.Routing
+{
+    internal class LeastConnectedLoadBalancingStrategy : ILoadBalancingStrategy
+    {
+        private readonly RoundRobinArrayIndex _readersIndex = new RoundRobinArrayIndex();
+        private readonly RoundRobinArrayIndex _writersIndex = new RoundRobinArrayIndex();
+
+        private readonly IClusterConnectionPool _connectionPool;
+        private readonly ILogger _logger;
+
+        public LeastConnectedLoadBalancingStrategy(IClusterConnectionPool connectionPool, ILogger logger)
+        {
+            _connectionPool = connectionPool;
+            _logger = logger;
+        }
+
+        public Uri SelectReader(IList<Uri> knownReaders)
+        {
+            return Select(knownReaders, _readersIndex, "reader");
+        }
+
+        public Uri SelectWriter(IList<Uri> knownWriters)
+        {
+            return Select(knownWriters, _writersIndex, "writer");
+        }
+
+        private Uri Select(IList<Uri> addresses, RoundRobinArrayIndex roundRobinIndex, string addressType)
+        {
+            var count = addresses.Count;
+            if (count == 0)
+            {
+                _logger.Trace($"Unable to select {addressType}, no known addresses given");
+                return null;
+            }
+
+            // choose start index for iteration in round-rodin fashion
+            var startIndex = roundRobinIndex.Next(count);
```

Comment:
The point here is to "randomize" address selection when all addresses have same amount of active connections. This is especially valuable when cluster has no active transactions. With simple for-loop we would always select the first available address. With round-robin selection of starting point we select machines in round-robin fashion. It could save us when accidentally fist machine is much worse than others.